### PR TITLE
Skip invalid socks in checkDispatchResult()

### PR DESCRIPTION
**Pull request by interma**
Port https://github.com/greenplum-db/gpdb/pull/14504 to main branch.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by adam8157** [comment]
Where does it set the `sock` to `-1`?

**Comment by adam8157** [comment]
@interma answered me offline, it's `pqFlush->pqSendSome->pqReadData->pqDropConnection`

**Review by adam8157 — APPROVED** [review]

File: src/backend/cdb/dispatcher/cdbdisp_async.c [hunk 1]
```diff
@@ -575,6 +575,48 @@ checkDispatchResultLoop(CdbDispatcherState *ds, int timeout_sec, WaitEventSet *w
 						 segdbDesc->whoami, PQerrorMessage(conn));
```
Comment (zxuejing) [inline src/backend/cdb/dispatcher/cdbdisp_async.c:575]: can we set dispatchResult->stillRunning = false here?

Comment (interma) [reply src/backend/cdb/dispatcher/cdbdisp_async.c:575]: I don't think so:
Set `stillRunning = false` will prevent all future works on the connection, but some errors may be recovery later, only the unrecovered connection (like sock==-1) need to set it.

**Pull request merged**